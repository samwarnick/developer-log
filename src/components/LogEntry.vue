<template>
  <li
    class="group flex justify-between items-start my-2 py-1 rounded-lg even:bg-grey-lightest odd:bg-grey-lighter text-grey-darkest"
  >
    <div class="flex-grow flex justify-between">
      <span class="px-4 font-semibold" v-html="convertToMarkdown(entry.content)"></span>
      <span class="text-grey-dark flex-no-shrink">{{ entry.created | moment("calendar") }}</span>
    </div>
    <button>
      <FontAwesomeIcon
        icon="trash-alt"
        class="text-grey-dark opacity-25 group-hover:opacity-100 hover:text-red-dark mx-2"
        @click="deleteLogEntry(entry.id)"
      ></FontAwesomeIcon>
    </button>
  </li>
</template>

<script>
import marked from "marked";
import * as logEntriesGql from "@/graphql/LogEntries.gql";
import * as deleteLogEntryGql from "@/graphql/DeleteLogEntry.gql";

export default {
  name: "LogEntry",
  props: {
    entry: Object
  },
  methods: {
    convertToMarkdown(value) {
      return marked(value, { sanitize: true });
    },
    deleteLogEntry(id) {
      this.$apollo.mutate({
        mutation: deleteLogEntryGql,
        variables: {
          input: { id }
        },
        update: store => {
          const data = store.readQuery({ query: logEntriesGql });

          data.logEntries = data.logEntries.filter(group => {
            group.logEntries = group.logEntries.filter(
              entry => entry.id !== id
            );
            return group.logEntries.length;
          });

          store.writeQuery({ query: logEntriesGql, data });
        },
        optimisticResponse: {
          __typename: "Mutation",
          deleteLogEntry: {
            __typename: "DeleteLogEntry",
            id: id
          }
        }
      });
    }
  }
};
</script>
