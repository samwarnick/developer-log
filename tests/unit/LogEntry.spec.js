import { shallowMount } from "@vue/test-utils";
import LogEntry from "@/components/LogEntry.vue";

const mutate = jest.fn();

describe("LogEntry.vue", () => {
  test("renders correctly", () => {
    const wrapper = shallowMount(LogEntry, {
      propsData: {
        entry: {
          id: "1234",
          content: "This is a test",
          created: "2018-11-10T21:39:43.115Z"
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test("should call apollo when deleting", () => {
    const wrapper = shallowMount(LogEntry, {
      propsData: {
        entry: {
          id: "1234",
          content: "This is a test",
          created: "2018-11-10T21:39:43.115Z"
        }
      },
      mocks: {
        $apollo: {
          mutate
        }
      }
    });
    wrapper.vm.deleteLogEntry("1234");
    expect(mutate).toBeCalled();
  });
});
