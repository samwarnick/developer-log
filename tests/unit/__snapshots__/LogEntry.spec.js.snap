// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogEntry.vue renders correctly 1`] = `
<li
  class="group flex justify-between items-start my-2 py-1 rounded-lg even:bg-grey-lightest odd:bg-grey-lighter text-grey-darkest"
>
  <div
    class="flex-grow flex justify-between"
  >
    <span
      class="px-4"
    >
      <p>
        This is a test
      </p>
      

    </span>
     
    <span
      class="text-grey-dark flex-no-shrink"
    >
      2018-11-10T21:39:43.115Z
    </span>
  </div>
   
  <button>
    <fontawesomeicon
      class="text-grey-dark opacity-25 group-hover:opacity-100 hover:text-red-dark mx-2"
      icon="trash-alt"
    />
  </button>
</li>
`;
